File: #1219.path-with-maximum-gold.cpp
class Solution {
public:
    int getMaximumGold(vector<vector<int>>& grid) {

        int result = 0;
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[0].size(); j++){
                if(grid[i][j]){
                    result = std::max(result, grid[i][j] + dfs(grid, i, j));
                }
            }
        }
        return result;
    }

private:
    vector<vector<int>> directions = {{1,0}, {-1,0}, {0,1}, {0,-1}};
    
    int dfs(vector<vector<int>>& grid, int i, int j){
        int ans = 0;
        int org = grid[i][j];
        grid[i][j] = 0;
        
        for(auto& direction : directions){
            int x = direction[0], y = direction[1];
            if(i + x >= 0 && j + y >= 0 && i + x < grid.size() 
               && j + y < grid[0].size()
               && grid[i+x][j+y] != 0) {
                ans = max(ans, grid[i+x][j+y] + dfs(grid, i + x, j + y));
            }
        }
        grid[i][j] = org;
        return ans;
    }
};

File: #1219.path-with-maximum-gold.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var getMaximumGold = function(grid) {
    function getMaximumGold(left, right) {
        if (left < 0 || right < 0 
            || left >= n1 
            || right >= n2 
            || !grid[left][right]) {
            return 0;
        }

        const arr1 = [1, -1, 0, 0];
        const arr2 = [0, 0, 1, -1];
        const cell = grid[left][right];
        let max = 0;
        grid[left][right] = 0;
        for (let i = 0; i < arr1.length; i++) {
            max = Math.max(max,
                getMaximumGold(left + arr1[i], right + arr2[i]));
        }
        grid[left][right] = cell;
        return max + cell;
    }

    const n1 = grid.length;
    const n2 = grid[0].length;
    let maxPath = 0;
    for (let i = 0; i < n1; i++) {
        for (let j = 0; j < n2; j++) {
            if (grid[i][j]) {
                maxPath = Math.max(maxPath, getMaximumGold(i, j));
            }
        }
    }
    return maxPath;
};
